ArjunaJTA/jta/tests/classes/com/hp/mwtests/ts/jta/xa/JTATest.java

```diff
@@ -342,6 +349,39 @@ public void rollback(Xid xid) throws XAException {
         }
     }
 
+    @Test
+    public void testRollbackCalledWhenEndHasXA_RBINTEGRITY() throws Exception {
+        jakarta.transaction.TransactionManager tm = com.arjuna.ats.jta.TransactionManager
+                .transactionManager();
+
+        tm.begin();
+
+        jakarta.transaction.Transaction theTransaction = tm.getTransaction();
+
+        assertTrue(theTransaction.enlistResource(new SimpleXAResource(){}));
+
+        assertTrue(theTransaction.enlistResource(new SimpleXAResource() {
+            @Override
+            public void end(Xid xid, int flags) throws XAException {
+                assertTrue(flags == XAResource.TMSUCCESS);
+                throw new XAException(XAException.XA_RBINTEGRITY);
+            }
+
+            @Override
+            public void rollback(Xid xid) throws XAException {
+                resource1Rollback = true;
+                throw new XAException(XAException.XAER_NOTA);
+            }
+        }));
+
+        try {
+            tm.commit();
+            fail("Should not have committed");
+        } catch (RollbackException e) {
+            assertTrue(resource1Rollback);
```

Comment:
WDYT about checking the XA errorCode as well?

Reply:
I'll answer here on behalf of this and the other similar thread that I would propose to keep for tracking that there is a question.

I am not sure which XA errorCode you are thinking of checking here, please can you expand on it?

Reply:
I didn't check if this is possible but I thought to extract the XA error code from the exception and assert that it was equal to the error code the resource threw

Reply:
I have added some tracking for what I can below, thanks for the suggestion